### DISCUSSION TITLE
Rounded borders

### DISCUSSION THREAD

**Discussion by voltumna** [fvwmorg/fvwm3#613] [General]
Hi All, 
I was wondering if there is a way to achieve rounded borders, I saw that there are a few patches for fvwm 2.6.9 [here](https://aur.archlinux.org/packages/mvwm-git) that provide this function, is this incorporated in this version? or is there a way to apply it?
I already 'solved' the translucency with compton, so this is the last piece of the puzzle

thank you

Maurizio

**Comment by voltumna** [comment]
Any one found a way to do this yet?

**Comment by ONykyf** [comment]
Would be nice to have. The last thing that stops me from switching to fvwm3.

**Comment by ONykyf** [comment]
A patch that adds RoundedCorners, RoundedCornersTop and RoundedCornersBottom with arbitrary radii for different corners (using Bresenham algorithm), together with SlightlyRoundedCorners (radius=3), is attached.
[fvwm3-1.1.2-roundedcorners.patch.gz](https://github.com/user-attachments/files/19778874/fvwm3-1.1.2-roundedcorners.patch.gz)

**Comment by ONykyf** [comment]
CornerLength (handle length) can also be different for all corners. Now menus and windows look like this.  Menus are set to `MenuStyle * RoundedCorners`, applications styles are configured to be `Style * RoundedCorners`, `HiddenHandles` as a default (xterm is shown), MidnightCommander has `RoundedCorners 0 22 0 15`,  `CornerLength 20 30 45 20`, `!HiddenHandles`, and xclock has `RoundedCorners 10 18 25 0` ,  `CornerLength 27`. The patch is a work in progress, but no obvious bugs encountered, hence is here just in case someone wants to try.
![screenshot](https://github.com/user-attachments/assets/da26d907-22b2-4898-96b4-7b3f1e8d62ce)
[fvwm3-1.1.2-roundedcorners.patch.gz](https://github.com/user-attachments/files/20421765/fvwm3-1.1.2-roundedcorners.patch.gz)

**Comment by ONykyf** [comment]
Style `RotateShadows` / `!RotateShadows` added. Неre is a Midnight Commander window with 

**Style BlueRoundedMC RoundedCorners 0 15 0 22, CornerLength 20 20 45 30
Style BlueRoundedMC TitleAtRight, RightTitleRotatedCW, !RotateShadows**
![notrotateshadows](https://github.com/user-attachments/assets/7d3268bc-1669-49c6-9bdf-85687646047a)

and with

**Style BlueRoundedMC RoundedCorners 0 15 0 22, CornerLength 20 20 45 30
Style BlueRoundedMC TitleAtRight, RightTitleRotatedCCW, RotateShadows**
![rotateshadows](https://github.com/user-attachments/assets/2f24a9d6-b282-4ef3-853a-4b08c7ea4548)

Updated patch against fvwm3-1.1.2:
[fvwm3-1.1.2-roundedcorners.patch.gz](https://github.com/user-attachments/files/20489867/fvwm3-1.1.2-roundedcorners.patch.gz)

**Comment by ONykyf** [comment]
As a grand rewrite of the decor code in the future is announced (although the time is unknown), and no major change to it will be accepted until then, let the code from a declined PR wait here until its time comes. I will update it with every new release of FVWM.

New features and styles are introduced, mainly to get rounded corners for window borders, buttons, menus etc. To achieve this, the code that draws relief rectangles has been heavily modified, mainly to provide correct junctions when two arms of an angle or two halves of an arc are drawn with different GCs. Bresenham algorithm is applied to construct arcs, both for drawing and for making windows shaped at corners using XShape (no antialising, hence some jagging is present, it is up to user whether to round corners or not).

Another distinctive feature is the possibility to specify different radii and corner lengths (handle lengths) for all corners of a window by `CornerLength` and `RoundedCorners` styles.

Decor elements are correctly rounded and rotated if `TitleAt{Top,Left,Bottom,Right}` with `XXXTitleRotated{CW,CCW}` is used. An option `RotateShadows` to rotate shadows together with title text is also introduced.

![screenshot-rounded](https://github.com/user-attachments/assets/2ab0bb4d-2da8-4e18-a80a-6f744d417335)

[fvwm3-1.1.2-roundedcorners.patch.gz](https://github.com/user-attachments/files/20532801/fvwm3-1.1.2-roundedcorners.patch.gz)

Comments, proposals and bug complaints are welcome!

**Comment by ONykyf** [comment]
Updated against fvwm-1.1.3. No problems observed yet.
[fvwm3-1.1.3-roundedcorners.patch.gz](https://github.com/user-attachments/files/20549883/fvwm3-1.1.3-roundedcorners.patch.gz)
